src/lib.rs

```diff
@@ -471,15 +437,20 @@ pub fn encode_without_checksum_to_fmt<T: AsRef<[u5]>>(
     hrp: &str,
     data: T,
 ) -> Result<fmt::Result, Error> {
-    let hrp_lower = match check_hrp(hrp)? {
+    let hrp = match check_hrp(hrp)? {
         Case::Upper => Cow::Owned(hrp.to_lowercase()),
         Case::Lower | Case::None => Cow::Borrowed(hrp),
     };
 
-    match Bech32WithoutChecksumWriter::new(&hrp_lower, fmt) {
-        Ok(mut writer) => Ok(writer.write(data.as_ref())),
-        Err(e) => Ok(Err(e)),
-    }
+    let mut write = || {
+        fmt.write_str(&hrp)?;
+        fmt.write_char(SEP)?;
+        for b in data.as_ref() {
+            fmt.write_char(b.to_char())?;
+        }
+        Ok(())
+    };
+    Ok(write())
```

Comment:
I'm a little uncertain about this. Seems to maintain the same behavior, but it is odd that an `fmt::Error` is always wrapped with `Ok`.

Reply:
Yeah, the existing API is really weird.

Reply:
Why are you constructing a closure and then immediately calling it?

Reply:
Mainly as a simple way to maintain the API behavior. i.e., wrapping a `Err(fmt::Error)` returned in each of the three places using `?` with an `Ok`.  Without the closure, it would be:

```rust
    if let Err(e) = fmt.write_str(&hrp) { return Ok(Err(e)); }
    if let Err(e) = fmt.write_char(SEP) { return Ok(Err(e)); }
    for b in data.as_ref() {
        if let Err(e) = fmt.write_char(b.to_char()) { return Ok(Err(e)); }
    }
    Ok(Ok(()))
```

Happy to update the error type is your prefer. It probably should be an enum that is either `bech32::Error` or `fmt::Error`. Or maybe adding a variant to `bech32::Error` for `fmt::Error`.

Reply:
This whole `Result<Result>` thing is weird to me -- I think we should add `fmt::Error` to the main `Error` enum and then flatten this return type.

I wouldn't worry about maintaining the API here. We want to overhaul this API anyway.

Reply:
Flatted the `Result`. Let me know if you prefer a different name for the variant or if you want to drop the wrapped `fmt::Error` as it currently doesn't provide any additional information.